docs/data/material/components/material-icons/SearchIcons.js

```diff
@@ -147,23 +154,61 @@ function handleLabelClick(event) {
   selectNode(event.currentTarget);
 }
 
+function isElmVisible(elm, margin = 0) {
+  const rect = elm.getBoundingClientRect();
+  return rect.bottom >= -margin && rect.top <= window.innerHeight + margin;
+}
+
 function Icon(props) {
-  const { icon, onOpenClick } = props;
+  const { icon, onOpenClick, initiallyVisible = false } = props;
+
+  const rootRef = React.useRef(null);
+  const [isVisible, setIsVisible] = React.useState(initiallyVisible);
+
+  // Virtualize the icons to reduce page size and React rendering time.
+  // Only render the icons after they become visible in the viewport.
+  React.useEffect(() => {
+    const margin = 200;
+    if (!rootRef.current) {
+      throw new Error('missing ref');
+    }
```

Comment:
Looking at the code, this case looks impossible. We can use `rootRef.current!`.

```suggestion
```

We could maybe do this: but `IntersectionObserver` seems cheap, so might just be extra complexity.

```suggestion
    if (initiallyVisible) {
      return;
    }
```

Reply:
Respectfully, I think we should avoid using non-null assertion operator. How do you feel about
```tsx
invariant(rootRef.current, 'Ref should be attached to a DOM node')
```
accessing refs in effects is a very common pattern and by using the non-null assertion operator, we're introducing a lot of type-unsafety.

Reply:
> How do you feel about

Invariant is even better, in dev mode 😁. What I'm pushing for is no defensive code that slows down the runtime for things that are guaranteed by React to never be needed. If we make a mistake, a user will report a bug, and we will fix it. We trade occasional mistakes for eventual better execution than alternatives.

Reply:
> What I'm pushing for is no defensive code that slows down the runtime for things that are guaranteed by React to never be needed

It's not guaranteed by React to never be needed. Someone 6 months from now, may decide to put a conditional around the element that holds the ref and the code breaks.

> We trade occasional mistakes for eventual better execution than alternatives.

imo this is a form of absolutism that doesn't lead to the best execution. Robustness and maintainability are part of the execution. Not having to investigate and open a bug is an end-user feature.
I agree that these concepts are sometimes in tension with raw performance. But an `invariant` call is a function call with an unentered conditional inside. And if you inline it you can even omit the function call.

Reply:
> It's not guaranteed by React to never be needed.

@Janpot Without the conditional logic, it's guaranteed to be defined through https://react.dev/learn/manipulating-the-dom-with-refs#when-react-attaches-the-refs.

With the conditional logic, than I agree, it's become about the tension around "trade occasional mistakes for eventual better execution than alternatives"

> Not having to investigate and open a bug is an end-user feature.

npm release tends to roll out at about 1% of the userbase per week. If the bug is rare, it means that less than 10% of the users will see it on that new version, so 0.1% total. Say this happens 4 times in a year. 0.4% of the user base. I think no defensive checks is worth a somewhat degraded DX for 0.4% of the user base 😄.

> invariant

I'm a bit concerned about the error message bundle size of this, we have so many ref accesses. Note that React removed the use of `invariant` with https://github.com/facebook/react/pull/22435.

Reply:
> Note that React removed the use of invariant with https://github.com/facebook/react/pull/22435.

The variants are still there, just inlined. I'm not advocating for an `invariant` function, I'm advocating for avoiding the non-null assertion operator. The trade-off can be negligable. With the new error infra it's trivial to even minify every error message by default.

My main question is just, do our users really perceive _this_ as one of the things that makes us bloated? I rather feel like the perceived bloat is in things like css-in-js overhead, supporting multiple styling runtimes, MD-specific UI like the ripple effect and its virtualization, features pulled in my bundles like e.g. pivoting, or excel export that my use-case doesn't really need, react 17 shims, cjs bundles that I don't want, typechecking overhead that our types add...

If users don't see an invariant check as bloat, then tackling it doesn't yield a win and all we're doing is make our code more brittle for contributors and maintainers.

Reply:
> do our users really perceive this as one of the things that makes us bloated

The value is likely about reducing a bit the bundlesize for end-users, and making the source a bit easier to follow. If it's helps with bundlephobia.com report value, great too.

About the feeling of bloat, I guess it's an emotional response for things people feels should be deleted.
Personally, this contributes a bit. I see the value in the check, but not enough to justify its costs, so it feels like something that shouldn't be here.

> If users don't see an invariant check as bloat, then tackling it doesn't yield a win and all we're doing is make our code more brittle for contributors and maintainers.

Agree with this framing, but also I think we should factor the bundle size and runtime cost of using this pattern everywhere. 

It feels like one of those cases: https://youtu.be/tRsxLLghL1k?si=sRhbVRHVLRR83pMX&t=574, be greedy, remove as much stuff as possible (not invariant, use !), then see if we can get away with it, or if reality forces us to add invariant checks back. So far, it seemed that ! was ok.